### PR TITLE
use git show instead of checking out files for file diff

### PR DESCRIPTION
**Pull request by jenna-a2ai**

### CODE REVIEW

**Comment by jenna-a2ai** [comment]
In this video you can see that:
- the metadata (modified time) for the resolved files don't change
- the git show files don't stay in the temp dir
- the file without an extension isn't copied with a "."
- you can repeatedly diff files in the app without resetting it

https://github.com/user-attachments/assets/0b5fbc57-b227-48c9-bbf9-f7ec119b793f

File: R/qc_fix_helper.R [hunk 1]
```diff
@@ -163,16 +163,38 @@ format_diff_section <- function(diff_lines) {
...
-  # create copy
-  copied_file <- name_file_copy(file_path)
-  file.copy(file_path, copied_file)
-  withr::defer_parent(
-    clean_up(file_path, copied_file)
-  )
+  temp_dir <- tempdir()
+  file_diff_dir <- file.path(temp_dir, "file_diff_dir")
+  fs::dir_create(file_diff_dir)
+  withr::defer_parent({
```
[the remark this answers is not in the document]
Comment (dpastoor) [reply R/qc_fix_helper.R:169]: i'm thinking of things like if get_script_contents is called multiple times by a parent - is the idea here to make it only clean up after the last one?

File: R/qc_fix_helper.R [hunk 2]
```diff
@@ -163,16 +163,38 @@ format_diff_section <- function(diff_lines) {
...
+    if (dir.exists(file_diff_dir)) {
+      fs::dir_delete(file_diff_dir)
+    }
+  }) #, envir = globalenv()
+
+  # name the files the contents will be redirected to
+  file_at_reference <- file.path(file_diff_dir, "reference")
```
Comment (dpastoor) [inline R/qc_fix_helper.R:176]: i would instead make these random file names, so they don't clash if this script is invoked multiple times. Then you don't need the subdir at all, and can just use tempfile() invocations

File: R/qc_fix_helper.R [hunk 3]
```diff
@@ -163,16 +163,38 @@ format_diff_section <- function(diff_lines) {
...
+  if (result_ref$status != 0) {
+    stop(result_ref$stderr)
+  }
+
+  # get reference file contents
+  command_comp <- glue::glue("git show {comparator}:{file_path} > {file_at_comparator}")
+  result_comp <- processx::run("sh", c("-c", command_comp), error_on_status = FALSE)
+
+  if (result_comp$status != 0) {
+    stop(result_comp$stderr)
```
Comment (dpastoor) [inline R/qc_fix_helper.R:192]: why just stderr

prefer to use rlang::abort

**Review by dpastoor — CHANGES_REQUESTED** [review]
some feedback!

File: R/qc_fix_helper.R [hunk 4]
```diff
@@ -163,16 +107,48 @@ format_diff_section <- function(diff_lines) {
...
-  # create copy
-  copied_file <- name_file_copy(file_path)
-  file.copy(file_path, copied_file)
-  withr::defer_parent(
-    clean_up(file_path, copied_file)
-  )
+  temp_dir <- tempdir()
+  file_diff_dir <- file.path(temp_dir, "file_diff_dir")
+  fs::dir_create(file_diff_dir)
+  withr::defer({
```
Comment (jenna-a2ai) [inline R/qc_fix_helper.R:113]: defer_parent was leftover from debugging, removed

File: R/qc_fix_helper.R [hunk 5]
```diff
@@ -163,16 +107,48 @@ format_diff_section <- function(diff_lines) {
...
+  command_ref <- glue::glue("git show {reference}:{file_path} > {file_at_reference}")
+  result_ref <- processx::run("sh", c("-c", command_ref), error_on_status = FALSE)
+
+  if (result_ref$status != 0) {
+    rlang::abort(message = glue::glue(
+    "status: {result_ref$status}
+    stdout: {result_ref$stdout}
+    stderr: {result_ref$stderr}
+    timeout: {result_ref$timeout}")
+    )
```
Comment (jenna-a2ai) [inline R/qc_fix_helper.R:133]: printing everything out from result now

Screenshot 2024-10-24 at 1 32 07 PM

File: R/qc_fix_helper.R [hunk 6]
```diff
@@ -163,16 +107,48 @@ format_diff_section <- function(diff_lines) {
...
+  command_comp <- glue::glue("git show {comparator}:{file_path} > {file_at_comparator}")
+  result_comp <- processx::run("sh", c("-c", command_comp), error_on_status = FALSE)
+
+  if (result_comp$status != 0) {
+    rlang::abort(message = glue::glue(
+      "status: {result_comp$status}
+      stdout: {result_comp$stdout}
+      stderr: {result_comp$stderr}
+      timeout: {result_comp$timeout}")
+    )
```
Comment (jenna-a2ai) [inline R/qc_fix_helper.R:146]: printing everything out from result now

File: R/qc_fix_helper.R [hunk 7]
```diff
@@ -163,16 +107,48 @@ format_diff_section <- function(diff_lines) {
...
+      fs::dir_delete(file_diff_dir)
+    }
+  })
 
-  # get file contents at the specified commits
-  reference_script <- read_file_at_commit(reference, file_path)
-  comparator_script <- read_file_at_commit(comparator, file_path)
+  # name the files the contents will be redirected to
+  file_at_reference <- tempfile(tmpdir = file_diff_dir)
+  file_at_comparator <- tempfile(tmpdir = file_diff_dir)
```
Comment (jenna-a2ai) [inline R/qc_fix_helper.R:121]: temp files instead of named files for no risk of repeated file names

File: R/qc_fix.R [hunk 8]
```diff
@@ -65,9 +65,13 @@ create_comment_body <- function(owner,
                                 comparator_commit = "current") {
 
   issue <- get_issue(owner, repo, issue_number)
+  ## check if file exists locally
+  if (!fs::file_exists(issue$title)) {
+    rlang::abort(glue::glue("Error: {issue$title} does not exist in local project repo"))
+  }
```
Comment (jenna-a2ai) [inline R/qc_fix.R:71]: Yikes this is sorta bad - I realized in implementing feedback that there's no error handling for when a file with an associated ghqc issue doesn't exist locally 

Screenshot 2024-10-24 at 1 34 45 PM

**Comment by jenna-a2ai** [comment]
fyi, there's this note about the license?

Screenshot 2024-10-24 at 1 39 41 PM

**Comment by jenna-a2ai** [comment]
In this video you can see that:
- the metadata (modified time) for the resolved files don't change
- a descriptive error occurs for deleted files


https://github.com/user-attachments/assets/a65bea9e-7dc2-4d01-826a-b41397ca6193

In this video you can see that:
- the git show files don't stay in the temp dir

https://github.com/user-attachments/assets/5de96c66-03c3-42db-921a-6b592e298a34

**Review by dpastoor — APPROVED** [review]
looks great thanks!

**Pull request merged**